File: python_src/new_interface.py
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import h5py
import facitpy

# ------------------------------
# Part 1: Inputs and profiles
# ------------------------------

nx = 100
nth = 100
nions = 5
Aa = 184

mp = 1.67262192369e-27
q_e = 1.602176634e-19
invaspct = 0.2
B0 = 3.7
R0 = 2.5
qa = 4
ma = Aa*mp

# radial and poloidal grids
xn = np.linspace(0, 0.99, nx)
theta = np.linspace(0, 2*np.pi, nth, endpoint=True)

# allocate arrays
Za = np.zeros(nx)
Ai = np.array([2,3,4,14,16])
Te = np.zeros(nx)
Ti = np.zeros(nx)
Ne = np.zeros(nx)
Na = np.zeros(nx)
Machi = np.zeros(nx)
gradTi = np.zeros(nx)
gradNa = np.zeros(nx)
qmag = np.zeros(nx)
dpsidx = np.zeros(nx)
FV = np.zeros(nx)
regulopt = np.array([0,0,0,0],dtype=np.float64)

BV = np.zeros((nx,nth))
RV = np.zeros((nx,nth))
Zi = np.zeros((nx,nions))
Ni = np.zeros((nx,nions))
gradNi = np.zeros((nx,nions))
jacob = np.zeros((nx,nth))
AsymPhi = np.zeros((nx,2))
AsymN = np.zeros((nx,nions,2))

term_iCRH = np.zeros(nx)
term_rot = np.zeros(nx)
delta_phi  = np.zeros(nx)
Deltam_phi  = np.zeros(nx)
S1 = np.zeros(nx)

pol_asym = True
rotation = True
solution = 3
full_geom = True

ni0 = [4e19,1e19,1e14,5e9,2e10]

# ------------------------------
# Densities
# ------------------------------
Ni[:,0] = ni0[0]*(1-xn**2)**2
Ni[:,1] = ni0[1]*(1-xn**2)**2
Ni[:,2] = ni0[2]*(1-xn**2)**2
Ni[:,3] = ni0[3]*(0.3+np.exp(6*(xn-1)))
Ni[:,4] = ni0[4]*(0.3+np.exp(6*(xn-1)))

Na = 6e12*(0.3+np.exp(4*(xn-1)))
Ne = 6e19*(1-xn**2)**2

Zi[:,0] = 1
Zi[:,1] = 1

for i in range(nx):
    Zi[i,2] = 1 if xn[i]>=0.6 else 2
    FV[i] = R0*B0

Zi[:,3] = np.clip(6 - 5*np.exp(-((xn-1)/0.25)**2), 1.0, 20.0)
Zi[:,4] = np.clip(7 - 6*np.exp(-((xn-1)/0.25)**2), 1.0, 20.0)
Za = np.clip(60 - 55*np.exp(-((xn-1)/0.25)**2), 1.0, 20.0)

qmag = 1 + (qa-1)*xn**2

# ------------------------------
# Temperaturas
# ------------------------------
T_floor = 0.05
Ti = 1.5*(1-xn**2)**2 + T_floor
Te = 3.0*(1-xn**2)**2 + 2.0*T_floor

# ------------------------------
# Geometry: RV, BV, jacob
# ------------------------------
min_denom = 1e-3
min_RV = 1e-2
min_qmag = 1e-6

for j in range(nth):
    denom = np.clip(1.0 - invaspct*np.cos(theta[j]), min_denom, None)

    RV[:, j] = R0*(1.0 + xn*invaspct*np.cos(theta[j]))
    RV[:, j] = np.clip(RV[:, j], min_RV, None)

    RV2 = RV[:, j]**2
    safe_RV2 = np.maximum(RV2, min_RV**2)
    safe_qmag2 = np.maximum(qmag**2, min_qmag**2)

    BV[:, j] = FV**2/safe_RV2 * (1.0 + invaspct**2/safe_RV2 * xn**2/safe_qmag2)

    jacob[:, j] = xn*R0/denom

# ------------------------------
# S1, S2, meff, anis, Omega, Machi
# ------------------------------
S1 = 0
S2 = Na.copy()
for j in range(nions):
    S1 += Ni[:,j]*Ai[j]*mp
    S2 += Ni[:,j]

meff = S1/np.maximum(S2,1e-12)
eps = xn*invaspct
anis = Te/Ti - 1.0
Omega = Za*q_e*B0/(Aa*mp)

with np.errstate(divide='ignore', invalid='ignore'):
    Machi_raw = (ma*Na + S1)*R0**2 * Omega/(S2*2.0*Ti)

Machi_raw = np.nan_to_num(Machi_raw)
Machi = np.clip(Machi_raw, 0.0, 0.8)

# ------------------------------
# dpsidx
# ------------------------------
for i in range(nx):
    dpsidx[i] = (2*xn[i]*qmag[i] - xn[i]**2 * (2*(4-1)*xn[i]))/max(qmag[i]**2,1e-12)
dpsidx[0] = dpsidx[1]

# ------------------------------
# Gradients
# ------------------------------
for i in range(nx):
    if i == 0:
        dx = xn[i+1] - xn[i]
        gradTi[i] = (Ti[i+1] - Ti[i]) / dx
        gradNa[i] = (Na[i+1] - Na[i]) / dx
        for j in range(nions):
            gradNi[i,j] = (Ni[i+1,j] - Ni[i,j]) / dx

    elif i == nx-1:
        dx = xn[i] - xn[i-1]
        gradTi[i] = (Ti[i] - Ti[i-1]) / dx
        gradNa[i] = (Na[i] - Na[i-1]) / dx
        for j in range(nions):
            gradNi[i,j] = (Ni[i,j] - Ni[i-1,j]) / dx

    else:
        dx = xn[i+1] - xn[i-1]
        gradTi[i] = (Ti[i+1] - Ti[i-1]) / dx
        gradNa[i] = (Na[i+1] - Na[i-1]) / dx
        for j in range(nions):
            gradNi[i,j] = (Ni[i+1,j] - Ni[i-1,j]) / dx

# ------------------------------
# Asymmetry
# ------------------------------
for i in range(nx):
    ratio_TeTi = max(Te[i]/Ti[i],1e-8)
    term_iCRH[i] = anis[i]/ratio_TeTi
    term_rot[i] = 2.0*np.sqrt(max(Machi[i],0.0))

    denom_phi = max(1.0 + Za[i]*Te[i]/Ti[i],1e-6)
    delta_phi[i] = eps[i]/denom_phi*(term_iCRH[i]+term_rot[i])
    Deltam_phi[i] = -0.165*eps[i]

AsymPhi[:,0] = delta_phi
AsymPhi[:,1] = Deltam_phi

for j in range(nions):
    AsymN[:,j,0] = -Zi[:,j]*(Te/Ti)*delta_phi
    AsymN[:,j,1] = -Zi[:,j]*(Te/Ti)*Deltam_phi

# ------------------------------
# Part 2: FACIT call
# ------------------------------
def arrF(a): return np.asfortranarray(np.asarray(a, dtype=np.float64))

raw_outputs = facitpy.facit_mod.facit(
    arrF(xn), arrF(theta), arrF(Za), float(Aa), arrF(Zi), arrF(Ai),
    arrF(Te), arrF(Ti), arrF(Ne), arrF(Ni), arrF(Na), arrF(Machi),
    arrF(gradTi), arrF(gradNi), arrF(gradNa),
    float(invaspct), float(B0), float(R0), arrF(qmag), arrF(dpsidx),
    arrF(FV), arrF(BV), arrF(RV), arrF(jacob),
    arrF(AsymPhi), arrF(AsymN),
    int(pol_asym), int(full_geom), int(solution), int(rotation),
    arrF(regulopt)
)

# ------------------------------
# Output structured
# ------------------------------
class FACITResult:
    def __init__(self, out):
        (
            self.Flux_imp, self.Da, self.Da_M,
            self.Vconv, self.Vconv_M, self.nn,
            self.dmin, self.dmaj,
            self.Da_PS_M, self.Da_BP_M, self.Da_CL_M,
            self.Ka_PS_M, self.Ka_BP_M, self.Ka_CL_M,
            self.Ha_PS_M, self.Ha_BP_M, self.Ha_CL_M,
            self.Va_PS_M, self.Va_BP_M, self.Va_CL_M,
            self.Da_PS, self.Da_BP, self.Da_CL,
            self.Ka_PS, self.Ka_BP, self.Ka_CL,
            self.Ha_PS, self.Ha_BP, self.Ha_CL,
            self.Va_PS, self.Va_BP, self.Va_CL
        ) = out

result = FACITResult(raw_outputs)

# ------------------------------
# Save EVERYTHING in HDF5
# ------------------------------
with h5py.File("resultado_facit_completo.h5","w") as f:

    # Inputs
    inputs = f.create_group("inputs")
    inputs["xn"] = xn
    inputs["theta"] = theta
    inputs["Za"] = Za
    inputs["Zi"] = Zi
    inputs["Ai"] = Ai
    inputs["Te"] = Te
    inputs["Ti"] = Ti
    inputs["Ne"] = Ne
    inputs["Ni"] = Ni
    inputs["Na"] = Na
    inputs["Machi"] = Machi
    inputs["gradTi"] = gradTi
    inputs["gradNi"] = gradNi
    inputs["gradNa"] = gradNa
    inputs["qmag"] = qmag
    inputs["dpsidx"] = dpsidx
    inputs["FV"] = FV
    inputs["BV"] = BV
    inputs["RV"] = RV
    inputs["jacob"] = jacob
    inputs["AsymPhi"] = AsymPhi
    inputs["AsymN"] = AsymN

    # Params
    params = f.create_group("params")
    params["Aa"] = Aa
    params["R0"] = R0
    params["B0"] = B0
    params["qa"] = qa
    params["invaspct"] = invaspct

    # Outputs
    outputs = f.create_group("outputs")
    for i, arr in enumerate(raw_outputs):
        outputs[f"out_{i}"] = np.asarray(arr)

    outputs["Da"] = result.Da
    outputs["Flux_imp"] = result.Flux_imp

print("✔ Todo guardado en resultado_facit_completo.h5 🚀")

# ------------------------------
# Plot
# ------------------------------
plt.figure()
plt.plot(xn, result.Da)
plt.title("Da(x)")
plt.grid()
plt.show()

File: fortran_src/new_interface.f90
program facit_try
    use facit_mod
    implicit none

    !===========================================================
    ! PARAMETERS
    !===========================================================
    integer :: nx, nth, nions, solution

    !===========================================================
    ! INPUT ARRAYS
    !===========================================================
    real(rkind), allocatable, dimension(:) :: theta
    real(rkind), allocatable, dimension(:) :: Te, Ti, Na, gradTi, Ne, gradNa
    real(rkind), allocatable, dimension(:,:) :: Ni, gradNi
    real(rkind), allocatable, dimension(:) :: qmag, xn, dpsidx, FV, Za
    real(rkind), allocatable, dimension(:) :: Machi2
    real(rkind), allocatable, dimension(:) :: Ai
    real(rkind), allocatable, dimension(:,:) :: Zi
    real(rkind), allocatable, dimension(:,:) :: jacob, BV,RV
    real(rkind), allocatable, dimension(:,:) :: AsymPhi
    real(rkind), allocatable, dimension(:,:,:) :: AsymN
    real(rkind), allocatable, dimension(:) :: regulopt
    logical  :: pol_asym, rotation, full_geom
    real(rkind)   :: B0, R0, invaspct, Aa

    !===========================================================
    ! OUTPUT ARRAYS
    !===========================================================

    real(rkind), allocatable, dimension(:) :: Da, Vconv, dmin, dmaj, Flux_imp
    real(rkind), allocatable, dimension(:,:) :: Da_M, Vconv_M
    real(rkind), allocatable, dimension(:,:) :: nn
    real(rkind), allocatable, dimension(:,:) ::  Da_BP_M, Da_PS_M, Da_CL_M, Ka_BP_M, Ka_PS_M, Ka_CL_M, Ha_BP_M, Ha_PS_M, Ha_CL_M, Va_BP_M, Va_PS_M, Va_CL_M
    real(rkind),  allocatable, dimension(:) :: Da_BP, Da_PS, Da_CL, Ka_BP, Ka_PS, Ka_CL, Ha_BP, Ha_PS, Ha_CL, Va_BP, Va_PS, Va_CL

    !===========================================================
    ! EXTRA ARRAYS
    !===========================================================

    real(rkind), allocatable, dimension(:) :: term_ICRH, term_rot, delta_phi, Deltam_phi, anis, epsi
    real(rkind), allocatable, dimension(:) :: S1, S2, Omega
    real(rkind)   :: ma
    integer :: i, j

    nx      = 100
    nth     = 100
    nions   = 5

    !===========================================================
    ! Allocate
    !===========================================================

    allocate(theta(nth),Te(nx), Ti(nx), Na(nx), gradTi(nx), Ne(nx), gradNa(nx))
    allocate(Ni(nx,nions), gradNi(nx,nions))
    allocate(qmag(nx), xn(nx), dpsidx(nx), FV(nx), Za(nx),Machi2(nx))
    allocate(Ai(nions))
    allocate(Zi(nx,nions))
    allocate(jacob(nx,nth), BV(nx,nth),RV(nx,nth))
    allocate(AsymPhi(nx,nth))
    allocate(AsymN(nx,nions,2))
    allocate(regulopt(4))
    allocate(Da(nx), Vconv(nx), dmin(nx), dmaj(nx), Flux_imp(nx))
    allocate(Da_M(nx,nions), Vconv_M(nx,nions))
    allocate(nn(nx,nth))
    allocate(Da_BP_M(nx,nions), Da_PS_M(nx,nions), Da_CL_M(nx,nions), Ka_BP_M(nx,nions), Ka_PS_M(nx,nions), Ka_CL_M(nx,nions), Ha_BP_M(nx,nions), Ha_PS_M(nx,nions), Ha_CL_M(nx,nions), Va_BP_M(nx,nions), Va_PS_M(nx,nions), Va_CL_M(nx,nions))
    allocate(Da_BP(nx), Da_PS(nx), Da_CL(nx), Ka_BP(nx), Ka_PS(nx), Ka_CL(nx), Ha_BP(nx), Ha_PS(nx), Ha_CL(nx), Va_BP(nx), Va_PS(nx), Va_CL(nx))
    allocate(term_ICRH(nx), term_rot(nx), delta_phi(nx), Deltam_phi(nx), anis(nx), epsi(nx))
    allocate(S1(nx), S2(nx), Omega(nx))

    !===========================================================
    ! Profiles
    !===========================================================

    invaspct = 0.2_rkind  ! Inverse aspect ratio 
    B0       = 3.7_rkind  ! Toroidal magnetic field on axis 
    R0       = 2.5_rkind  ! Major radius
    Aa       = 184.0_rkind ! Atomic mass number of the main impurity
    ma       = Aa*mp  

    do i = 1, nx
        xn(i) = 0.01_rkind*(i-1)
    end do

    do i = 1, nth
        theta(i) = (i-1) * 2.0_rkind*pi / (nth-1)
    end do

    Ai = (/2._rkind, 3._rkind, 4._rkind, 14._rkind, 16._rkind/)

    regulopt = 0.0_rkind

    do i=1,nx
        ! Ion densities
        Ni(i,1) = 4e19_rkind*(1.0_rkind-xn(i)**2)**2  ! Deuterium (example)
        Ni(i,2) = 1e19_rkind*(1.0_rkind-xn(i)**2)**2  ! Tritium (example)
        Ni(i,3) = 1e14_rkind*(1.0_rkind-xn(i)**2)**2  ! Helium (example)
        Ni(i,4) = 5e9_rkind*(0.3_rkind + exp(6.0_rkind*(xn(i)-1.0_rkind))) ! Minor impurity 1 (edge peaked)
        Ni(i,5) = 2e10_rkind*(0.3_rkind + exp(6.0_rkind*(xn(i)-1.0_rkind))) ! Minor impurity 2 (edge peaked)
        Na(i)   = 6e12_rkind*(0.3_rkind + exp(6.0_rkind*(xn(i)-1.0_rkind))) ! Main impurity density (edge peaked)
        Ne(i)   = 6e19_rkind*(1.0_rkind-xn(i)**2)**2  ! Electron density (charge neutrality is not strictly enforced here for simplicity)

        ! Ion charges
        Zi(i,1)=1.0_rkind
        Zi(i,2)=1.0_rkind
        if (xn(i)>=0.6_rkind) then
            Zi(i,3)=1.0_rkind
        else
            Zi(i,3)=2.0_rkind
        end if
        Zi(i,4)= 6.0_rkind - 5.0_rkind*exp(-((xn(i)-1.0_rkind)/0.25_rkind)**2)
        Zi(i,5)= 7.0_rkind - 6.0_rkind*exp(-((xn(i)-1.0_rkind)/0.25_rkind)**2)
        Za(i)= 60.0_rkind - 55.0_rkind*exp(-((xn(i)-1.0_rkind)/0.25_rkind)**2) ! Main impurity charge

        ! Temperatures 
        Ti(i) = 1.5_rkind*(1.0_rkind-xn(i)**2)**2
        Te(i) = 3.0_rkind*(1.0_rkind-xn(i)**2)**2

       
        qmag(i) = 1.0_rkind + (4-1.0_rkind)*xn(i)**2
        FV(i) = R0*B0
    enddo

    do i = 1,nx  
        do j = 1, nth
            RV(i,j) = R0 * (1.0_rkind + xn(i)*invaspct*cos(theta(j))) 
            BV(i,j) = FV(i)**2 / (RV(i,j)**2) * (1.0_rkind + invaspct**2 /(RV(i,j)**2) * xn(i)**2 /(qmag(i)**2) ) 
            jacob(i,j) = xn(i) * R0 / (1.0_rkind - invaspct*cos(theta(j)))
        enddo
        epsi(i) = xn(i) * invaspct 
        anis(i) = (Te(i)/Ti(i) - 1.0_rkind) 
    enddo

    S1(:) = 0.0_rkind 
    S2(:) = Na(:) 

    do j = 1, nions
        S1 = S1 + Ni(:,j)*Ai(j) *mp 
        S2 = S2 + Ni(:,j)
    end do

    ! Impurity rotation frequency (Omega)
    Omega = Za * q_e * B0/(Aa *mp ) 
    
    ! Mach number squared (simplified)
    Machi2 = (ma*Na + S1)* R0**2 *Omega/(S2*2.0_rkind*Ti)

    ! Calculate dpsidx (derivative of poloidal flux with respect to radial coordinate)
    do i = 1, nx
        dpsidx(i) = ( 2.0_rkind * xn(i) * qmag(i) - xn(i)**2 * (2.0_rkind * (4-1.0_rkind) * xn(i)) )  / ( qmag(i)**2 )
    end do

    ! Calculate Asymmetries (AsymPhi and AsymN)
    do i=1,nx 
        term_ICRH(i) = anis(i)  / ( (Te(i)/Ti(i))) 
        term_rot(i)  = sqrt(Machi2(i)) * 2.0_rkind 
        
        delta_phi(i) = epsi(i)/(1.0_rkind + Za(i)*Te(i)/Ti(i)) * (term_ICRH(i) + term_rot(i))
        Deltam_phi(i) = -0.165_rkind*epsi(i)        
    enddo
      
    AsymPhi(:,1) = delta_phi
    AsymPhi(:,2) = Deltam_phi

    ! Density asymmetries (simplified relation to potential)
    do j = 1, nions
    !    AsymN(:,j,1) = - Zi(:,j) * (Te/Ti) * delta_phi
    !    AsymN(:,j,2) = - Zi(:,j) * (Te/Ti) * Deltam_phi
        AsymN(:,j,1) = 1
        AsymN(:,j,2) = 1
    enddo


    !===========================================================
    ! CALCULATE GRADIENTS (Centered finite difference)
    !===========================================================
    ! Grad Ni
    do j= 1, nions
        gradNi(1,j) = (Ni(2,j) - Ni(1,j)) / (xn(2) - xn(1) +1e-33)
        do i = 2, nx-1
            gradNi(i,j) = (Ni(i+1,j) - Ni(i-1,j)) / (xn(i+1) - xn(i-1)+1e-33)
        end do
        gradNi(nx,j) = (Ni(nx,j) - Ni(nx-1,j)) / (xn(nx) - xn(nx-1)+1e-33)
    enddo

    ! Grad Na
    gradNa(1) = (Na(2) - Na(1)) / (xn(2) - xn(1)+1e-33)
    do i = 2, nx-1
        gradNa(i) = (Na(i+1) - Na(i-1)) / (xn(i+1) - xn(i-1)+1e-33)
    end do
    gradNa(nx) = (Na(nx) - Na(nx-1)) / (xn(nx) - xn(nx-1)+1e-33)

    ! Grad Ti
    gradTi(1) = (Ti(2) - Ti(1)) / (xn(2) - xn(1)+1e-33)
    do i = 2, nx-1
        gradTi(i) = (Ti(i+1) - Ti(i-1)) / (xn(i+1) - xn(i-1)+1e-33)
    end do
    gradTi(nx) = (Ti(nx) - Ti(nx-1)) / (xn(nx) - xn(nx-1)+1e-33)


    !===========================================================
    ! SET OPTIONS (These are now module variables)
    !===========================================================
    pol_asym    = .true.
    rotation    = .true.
    full_geom   = .false.
    solution    = 3

    call FACIT(nx, nth, xn,nions, theta, Za, Aa, Zi, Ai, Te, Ti, Ne, Ni, Na, Machi2, &
                    gradTi, gradNi, gradNa, invaspct, B0, R0, qmag, dpsidx, FV, BV, RV, jacob, AsymPhi, AsymN, & 
                    pol_asym, full_geom,solution, rotation, regulopt, Flux_imp, Da, Da_M, Vconv,Vconv_M, nn, dmin, dmaj, & 
                    Da_PS_M, Da_BP_M, Da_CL_M, Ka_PS_M, Ka_BP_M, Ka_CL_M, Ha_PS_M, Ha_BP_M, Ha_CL_M,Va_PS_M,Va_BP_M,Va_CL_M, &
                    Da_PS, Da_BP, Da_CL, Ka_PS, Ka_BP, Ka_CL, Ha_PS, Ha_BP, Ha_CL,Va_PS,Va_BP,Va_CL) 

     
     
    open(20,file='facit_output.dat')
    do i=1,nx
        write(20,203) xn(i), Da(i), Vconv(i), Flux_imp(i), dmin(i), dmaj(i)
    enddo
    close(20)

    203 format(6(E14.7,1X))

    deallocate(theta,Te, Ti, Na, gradTi, Ne, gradNa)
    deallocate(Ni, gradNi)
    deallocate(qmag, xn, dpsidx, FV, Za,Machi2)
    deallocate(Ai)
    deallocate(Zi)
    deallocate(jacob, BV,RV)
    deallocate(AsymPhi)
    deallocate(AsymN)
    deallocate(regulopt)
    deallocate(Da, Vconv, dmin, dmaj, Flux_imp)
    deallocate(Da_M, Vconv_M)
    deallocate(nn)
    deallocate(Da_BP_M, Da_PS_M, Da_CL_M, Ka_BP_M, Ka_PS_M, Ka_CL_M, Ha_BP_M, Ha_PS_M, Ha_CL_M, Va_BP_M, Va_PS_M, Va_CL_M)
    deallocate(Da_BP, Da_PS, Da_CL, Ka_BP, Ka_PS, Ka_CL, Ha_BP, Ha_PS, Ha_CL, Va_BP, Va_PS, Va_CL)
    deallocate(term_ICRH, term_rot, delta_phi, Deltam_phi, anis, epsi)
    deallocate(S1, S2, Omega)

end program facit_try
